### PR TITLE
Linter and Formatter added

### PR DESCRIPTION
**Pull request by mohammedmustafaahmed**
Refer to #46 

Assalam Walaykum. This commit adds [flake8](https://flake8.pycqa.org/en/latest/) linter and [black](https://black.readthedocs.io/en/stable/) formatter. This also actually runs the formatter and linter against the python codes with some adjustments.

At the moment there is no pre-commit hook set up, but inshallah if the owners prefer it, I can add another pull request for it with updated instructions (refer to [this](https://github.com/sunnah-com/api/issues/46#issuecomment-651585379))

### How to test this commit

```sh
# check out to this pull request (or checkout to the branch)
source ./venv/bin/activate
pip3 install -r requirements.txt
black .
flake8 .
```
note that you will still see some linting errors, since I have set the line length to be 88 and some strings are just too large.

### Test application:

```sh
docker-compose up
# visit localhost:5000, localhost:5000/v1/collections/ 
```

image

### CODE REVIEW

**Comment by ahadith** [comment]
Note for @Glaisher when he reviews this - one of the commits pulls forward the separate import line but a subsequent commit puts it back where it was.

**Comment by ahadith** [comment]
Jazakallahu khair @mohammedmustafaahmed - I love the initiative you took. We will review this and get back with comments or merge it as soon as we can in sha Allah!

**Comment by mohammedmustafaahmed** [comment]
> Jazakallahu khair @mohammedmustafaahmed - I love the initiative you took. We will review this and get back with comments or merge it as soon as we can in sha Allah!

Inshallah :)

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[flake8]
+ignore = F403
```
Comment (Glaisher) [inline .flake8:2]: Why is this needed? We don't have any usage of `from module import *` afaict

File: main.py [hunk 2]
```diff
@@ -5,103 +5,135 @@
...
         result = result.serialize()
         return jsonify(result)
+
     return decorated_function
 
-@app.route('/', methods=['GET'])
+
+@app.route("/", methods=["GET"])
 def home():
     return "<h1>Welcome to sunnah.com API.</p>"
```
Comment (Glaisher) [inline main.py:59]: Can we fix this in this commit as well? Not really python issue but it's a lint error nonetheless :)

File: main.py [hunk 3]
```diff
@@ -5,103 +5,135 @@
...
+@app.route("/v1/collections/<string:name>/books/<string:bookNumber>", methods=["GET"])
 @single_resource
 def api_collection_book(name, bookNumber):
     book_id = Book.get_id_from_number(bookNumber)
     return Book.query.filter_by(collection=name, ourBookID=book_id)
 
-@app.route('/v1/collections/<string:collection_name>/books/<string:bookNumber>/hadiths', methods=['GET'])
+
+@app.route(
+    "/v1/collections/<string:collection_name>/books/<string:bookNumber>/hadiths",
```
Comment (Glaisher) [inline main.py:88]: I'd suggest increasing the max line length instead of breaking into different lines in some `app.route` registrations and using single line in others? It would be better to have it consistent throughout the file.
 Also max line length is pretty controversial :) !

File: .flake8 [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+max-line-length = 88
+max-complexity = 10
+exclude =
```
Comment (Glaisher) [inline .flake8:5]: Just a suggestion. Instead of adding exclusions, can we move all the python source files to a subdirectory (say `sunnah_api/` for example)? Our root directory is now pretty crowded so this might be a good opportunity to do this.
We can maybe do this in a later PR too if it's too complicated.

Comment (mohammedmustafaahmed) [reply .flake8:5]: I think its better to have 2 separate PRs, i make another PR for that inshallah!

File: main.py (hunk 3 above)
Comment (mohammedmustafaahmed) [reply main.py:88]: Good idea, ill use Django's default 119

File: .flake8 (hunk 1 above)
Comment (mohammedmustafaahmed) [reply .flake8:2]: Oops will remove it

File: main.py (hunk 2 above)
Comment (mohammedmustafaahmed) [reply main.py:59]: Sorry i was just confused which line you were referring to? This just adds a new line after `return decorated_function` ?

File: main.py (hunk 3 above)
Comment (mohammedmustafaahmed) [reply main.py:88]: a warning though that this might compress dictionary representations too especially in the `serialise()` methods of `models.py` @Glaisher

**Comment by mohammedmustafaahmed** [comment]
@Glaisher inshallah let me know what you think now, i tried to address all your comments :)

File: main.py (hunk 2 above)
Comment (Glaisher) [reply main.py:57]: Sorry I meant the html part which has mismatched tags.

File: main.py (hunk 3 above)
Comment (Glaisher) [reply main.py]: > a warning though that this might compress dictionary representations too especially in the `serialise()` methods of `models.py` @Glaisher

Oh is that done by black autoformatting? I guess it's not much of an issue. :)

File: main.py (hunk 2 above)
Comment (mohammedmustafaahmed) [reply main.py]: Oh sorry i overlooked that haha! Fixed it :)

**Comment by mohammedmustafaahmed** [comment]
@Glaisher any comments on the fixed changes?

File: main.py [hunk 5]
```diff
@@ -5,103 +5,115 @@
 
 
 app = Flask(__name__)
-app.config.from_object('config.Config')
+app.config.from_object("config.Config")
+
+from models import HadithCollection, Book, Chapter, Hadith
```
Comment (Glaisher) [inline main.py:10]: This line requires a flake ignore for E402 rule (module level import not at top of file)

**Comment by mohammedmustafaahmed** [comment]
@Glaisher yes i forgot about the readme, thanks for the reminder. Should be updated now :)

**Comment by ahadith** [comment]
Jazakallahu khair @mohammedmustafaahmed !

**Comment by ahadith** [comment]
@mohammedmustafaahmed would you send us a contact message on the website with your email address please?

**Comment by mohammedmustafaahmed** [comment]
> @mohammedmustafaahmed would you send us a contact message on the website with your email address please?

@ahadith Yes sure. Sent a message already :) Let me know if you received it inshallah.
